MINIFICPP-1797 Python bootstrap (part 1)

martinzink opened the pull request:
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

szaszm commented on CMakeLists.txt:
```diff
@@ -269,13 +260,12 @@ add_library(ut INTERFACE)
 
 # cURL
-if(NOT DISABLE_CURL)
+if(ENABLE_CURL)
```
It doesn't strictly belong here, but I wouldn't be against hardcoding these to true. The usability of minifi is very limited without curl and openssl.

martinzink replied:
Good idea, i've created a ticket for this. https://issues.apache.org/jira/browse/MINIFICPP-2251

szaszm reviewed (commented):
How can we test this? Could you update the README with instructions?

If it's ready on Linux, bootstrap.sh and related scripts can also be removed.

martinzink commented on .gitignore:
```diff
@@ -48,6 +48,7 @@ cmake-build-debug
 *flowfile_checkpoint
 build
 /*build*
+!win_build_vs.bat
```
Yes, the one above it. `/*build*` matches `win_build_vs.bat` since its in the root directory

szaszm commented on README.md:
````diff
@@ -304,8 +304,30 @@ sudo brew install libpcap
+
+Set up a virtual environment in the bootstrap folder.
+```
+python -m venv venv
````
```suggestion
cd bootstrap
python -m venv venv
```

martinzink replied:
I've added wrapper scripts, so I've rephrased this section. https://github.com/apache/nifi-minifi-cpp/pull/1681/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R309

szaszm commented on README.md:
````diff
@@ -304,8 +304,30 @@ sudo brew install libpcap
+pip install -r requirements.txt
+```
+
+After that you can run the bootstrap script that will guide you through the configuration, dependency installation and build processes.
````
```suggestion
After that you can run the bootstrap script: `python main.py`. It will guide you through the configuration, dependency installation and build processes.
```

martinzink replied:
I've added wrapper scripts, so I've rephrased this section. https://github.com/apache/nifi-minifi-cpp/pull/1681/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R309

szaszm commented on bootstrap/system_dependency.py:
```diff
@@ -0,0 +1,168 @@
+
+
+def install_compiler(no_confirm: bool) -> str:
+    print("For CMake to work we need a working compiler")
```
This should check first whether a compiler is already available on the system.

martinzink replied:
Now it only installs these packages if they are missing. You can also skip the installation of the default compiler with a flag aswell

szaszm commented on bootstrap/cli.py:
```diff
@@ -0,0 +1,141 @@
+    return False
+
+
+def bool_menu(minifi_options: MinifiOptions) -> bool:
+    possible_values = [option_name for option_name in minifi_options.bool_options]
+    selected_values = [option.name for option in minifi_options.bool_options.values() if option.value == "ON"]
+    questions = [
+        inquirer.Checkbox(
+            "options",
+            message="MiNiFi C++ Options",
+            choices=possible_values,
+            default=selected_values
+        ),
+    ]
+
+    answers = inquirer.prompt(questions)
+    for bool_option in minifi_options.bool_options.values():
+        if bool_option.name in answers["options"]:
+            bool_option.value = "ON"
+        else:
+            bool_option.value = "OFF"
+
+    return False
```
This menu needs guidance on how to exit from it after selecting the options.
![image](https://github.com/apache/nifi-minifi-cpp/assets/1170582/041bf975-0bcd-4b95-9dfd-b46f2915136d)

martinzink replied:
I've added some clarifications to the title of this menu.
https://github.com/apache/nifi-minifi-cpp/pull/1681/files#diff-540b6c7fb2ac6e2c1e8c04096674ab473e4e16eaf2f06cff971b097ae65bc2cdR104

szaszm commented on bootstrap/system_dependency.py:
```diff
@@ -0,0 +1,168 @@
+        system_dependencies['python'] = 'python'
+    if minifi_options.is_enabled("MINIFI_OPENSSL"):
+        system_dependencies['openssl'] = 'perl'
+    if minifi_options.is_enabled("ENABLE_JNI"):
+        system_dependencies['jni'] = ''
```
This is missing.
```
CMake Error at bootstrap/venv/lib/python3.10/site-packages/cmake/data/share/cmake-3.26/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find JNI (missing: JAVA_INCLUDE_PATH JAVA_INCLUDE_PATH2 AWT)
```

martinzink replied:
my bad, I've added the missig dependencies 👍

szaszm commented on bootstrap/system_dependency.py:
```diff
@@ -0,0 +1,168 @@
+    assert _run_command_with_confirm(command, no_confirm)
+
+
+def _install(dependencies_str: str, no_confirm: bool):
```
Only the missing dependencies should be installed. There are ways to query already installed packages with all package managers.

martinzink replied:
Good idea, I've refactored the whole thing around this idea

szaszm reviewed (commented):
after testing on ubuntu

szaszm commented on bootstrap/cli.py:
```diff
@@ -0,0 +1,141 @@
+    if not os.path.exists(minifi_options.build_dir):
+        os.mkdir(minifi_options.build_dir)
+    os.chdir(minifi_options.build_dir)
+    cmake_cmd = f"cmake -G Ninja {minifi_options.create_cmake_options_str()} {minifi_options.source_dir}"
```
This could default to `FORCE_COLORED_OUTPUT=ON`, since Ninja buffers the output instead of just letting the subprocesses write to the terminal directly, resulting in no color.

szaszm replied:
We could do that. Automations could still explicitly set it to off if needed. We should mention this in the readme if we change the default.

martinzink replied:
Changed the default, https://github.com/apache/nifi-minifi-cpp/pull/1681/commits/0e43424c2a2f2567876534b06fa68ef005104874
But im not sure where (if at all) we should include this in the readme

szaszm commented on bootstrap/system_dependency.py:
```diff
@@ -0,0 +1,168 @@
+    print("For CMake to work we need a working compiler")
+    platform_system = platform.system()
+    if platform_system == "Darwin":
+        _install("llvm", no_confirm)
```
Couldn't this use the existing XCode AppleClang?
I tried this on MacOS, and it defaulted to trying to compile with GCC (installed previously manually with Homebrew) after the generation, despite the XCode tools being installed. It doesn't work at the moment, but we should be able to change the preferred compiler if multiple are available/installed.

martinzink replied:
Could you recheck it now after the refactor?

szaszm replied:
This problem stemmed from an environment issue on my side, but a bootstrap issue.

szaszm reviewed (commented):
tried on macos

szaszm commented:
Got this on Windows, after I managed to install winget and accept the source terms by running it once  manually:
```
(venv) C:\a\m\bootstrap>python main.py
For CMake to work we need a working compiler
Running winget install --disable-interactivity --accept-package-agreements msvc [y/n]
y
No package found matching input criteria.
Traceback (most recent call last):
  File "C:\a\m\bootstrap\main.py", line 18, in <module>
    compiler_override = install_compiler(no_confirm)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\a\m\bootstrap\system_dependency.py", line 159, in install_compiler
    _install("msvc", no_confirm)
  File "C:\a\m\bootstrap\system_dependency.py", line 115, in _install
    _install_with_winget(dependencies_str, no_confirm)
  File "C:\a\m\bootstrap\system_dependency.py", line 97, in _install_with_winget
    assert _run_command_with_confirm(command, no_confirm)
AssertionError
```

martinzink commented on README.md:
```diff
@@ -304,8 +304,30 @@ sudo brew install libpcap
+
+#### Python based bootstrapping
+
+Set up a virtual environment in the bootstrap folder.
```
Good idea, I've created a wrapper https://github.com/apache/nifi-minifi-cpp/blob/02abe24dd97f8a160fd1fe0e2982e27cb7e5936e/bootstrap/py_bootstrap.bat https://github.com/apache/nifi-minifi-cpp/blob/02abe24dd97f8a160fd1fe0e2982e27cb7e5936e/bootstrap/py_bootstrap.sh

martinzink commented on bootstrap/cli.py:
```diff
@@ -0,0 +1,141 @@
+        main_menu_options = {
+            f"Build dir: {minifi_options.build_dir}": build_dir_menu,
+            f"Build type: {minifi_options.build_type.value}": build_type_menu,
+            "Build options": bool_menu,
```
Good idea, that would be a good way to improve it in a followup PR

martinzink commented on bootstrap/system_dependency.py:
```diff
@@ -0,0 +1,168 @@
+
+    command = "sudo apt install -y {}".format(
+        _replace_wholewords(dependencies_str, replace_dict))
+    assert _run_command_with_confirm(command, no_confirm)
```
You are right, I've removed the assertions from these places

martinzink commented on bootstrap/cli.py:
```diff
@@ -0,0 +1,141 @@
+    return False
+
+
+def bool_menu(minifi_options: MinifiOptions) -> bool:
```
I think that would be possible, but since its not trivial I think I'd rather do that in part 2 aswell.

martinzink commented on bootstrap/cli.py:
```diff
@@ -0,0 +1,141 @@
+            f"Build dir: {minifi_options.build_dir}": build_dir_menu,
+            f"Build type: {minifi_options.build_type.value}": build_type_menu,
+            "Build options": bool_menu,
+            "One click build": do_one_click_build,
```
I've added a feature so it only installs if there is some dependency missing, so it wont be annoying, would that suffice?

martinzink commented on bootstrap/system_dependency.py:
```diff
@@ -0,0 +1,168 @@
+from minifi_option import MinifiOptions
+
+
+def _query_yes_no(question: str, no_confirm: bool) -> bool:
```
Now that we dont reinstall the packages, one would only encounter this a limited amount of times, so this might be not neccessary. What do you think?

lordgamez replied:
If it's minimal now I think that's okay

martinzink commented:
I've refactored pretty much the whole thing. Now it only installs the packages if they are not installed.
Rewrote the CI to use the new bootstrapping process.

On windows it only supports chocolatey, due to various problems with winget

szaszm commented on bootstrap/system_dependency.py:
```diff
@@ -0,0 +1,55 @@
+    if minifi_options.is_enabled("ENABLE_PYTHON_SCRIPTING"):
+        system_dependencies['python'] = {'python'}
+    if minifi_options.is_enabled("MINIFI_OPENSSL"):
+        system_dependencies['openssl'] = {'perl'}
```
OpenSSL also needs NASM, at least on Windows. Is it handled elsewhere?

martinzink replied:
Thats packaged with strawberryperl, (with the whole toolchain but that causes other problems). Usually we wipe the c:\Strawberry\c\bin dir to fix those problems, but then we also wipe the NASM which is required, so we had to install that aswell, instead Im deleting the installed toolchain but leave NASM alone. 
https://github.com/martinzink/nifi-minifi-cpp/blob/python_bootstrap/bootstrap/package_manager.py#L178

(installing NASM separatly came with its own issues, since neither choco nor winget sets the PATH for NASM)

szaszm replied:
Wiping the Strawberry Perl bin directory is a bit too extreme IMO, possibly ruining the user's existing installation. Would it be possible to only fix the PATH envvar instead? Like maybe moving the Strawberry Perl bin directory to the end of it, so the normal patch command is found first.

martinzink replied:
I replaced this approach with only temporarly changing the path and not deleting anything.
(I've also added NASM as a separate dependency)
https://github.com/apache/nifi-minifi-cpp/pull/1681/commits/e611b1e44e82a01e082c3b8a1099d2d7591d7401#diff-0ebcdc25be3be8a77e8b404eb6962891695836a8ca6bfa65607a2622d114f35dR235-R237

szaszm commented on bootstrap/py_bootstrap.bat:
```diff
@@ -0,0 +1,15 @@
+@echo off
+
+set SCRIPT_DIR=%~dp0
+set VENV_DIR=%SCRIPT_DIR%venv
+
+if exist %VENV_DIR% (
+    call %VENV_DIR%\Scripts\activate
+) else (
+    echo Creating virtualenv
+    python -m venv %VENV_DIR%
+    call %VENV_DIR%\Scripts\activate
+    pip install -r %SCRIPT_DIR%requirements.txt
+)
+
+python %SCRIPT_DIR%main.py
```
I like to use quotes, even when they're not strictly required. Windows cmd scripts have a strange syntax, so you usually need to quote the whole argument, not just the part containing the variable reference, otherwise the quotes just become part of the argument.
```suggestion
set "SCRIPT_DIR=%~dp0"
set "VENV_DIR=%SCRIPT_DIR%venv"

if exist "%VENV_DIR%" (
    call "%VENV_DIR%\Scripts\activate"
) else (
    echo Creating virtualenv
    python -m venv "%VENV_DIR%"
    call "%VENV_DIR%\Scripts\activate"
    pip install -r "%SCRIPT_DIR%requirements.txt"
)

python "%SCRIPT_DIR%main.py"
```

szaszm commented on bootstrap/package_manager.py:
```diff
@@ -0,0 +1,325 @@
+
+    def install_compiler(self) -> str:
+        if distro.id() == "ubuntu" and int(distro.major_version()) < 22:
+            self.install({"compiler_prereq": {"software-properties-common"}})
```
Most 3rd party repos also need: `apt-transport-https ca-certificates software-properties-common`

It may be more noticable with a minimal server install or docker image, but I didn't test yet.

martinzink replied:
👍 https://github.com/apache/nifi-minifi-cpp/pull/1681/commits/6b7d1af1293bf49e121d32f90e6cd9617c78a68e#diff-0ebcdc25be3be8a77e8b404eb6962891695836a8ca6bfa65607a2622d114f35dR125

szaszm commented on bootstrap/package_manager.py:
```diff
@@ -0,0 +1,325 @@
+                             replace_dict={"patch": set(),
+                                           "jni": {"maven"}})
+
+    def install_compiler(self) -> str:
+        self.install({"compiler": {"llvm"}})
+        return ""
```
Can this use AppleClang from XCode on Mac?

martinzink replied:
Sure, I was confused but as it turns out xcode should be installed already if the user has access to brew. So we dont need to install anything here. It should use the default compiler (what a hello world project would pick with cmake). Can you check?

szaszm replied:
I checked last week, and it was working as well as without the new python bootstrap. (Which still doesn't build on my mac, but that's not your fault.)

szaszm commented on bootstrap/main.py:
```diff
@@ -0,0 +1,56 @@
+        parser = argparse.ArgumentParser()
+        parser.add_argument('--noconfirm', action="store_true", default=False,
+                            help="Bypass any and all “Are you sure?” messages.")
+        parser.add_argument('--minifi_options', default="", help="Overrides the default minifi options during the "
+                                                                 "initial parsing")
+        parser.add_argument('--cmake_options', default="", help="Appends this to the final cmake command")
+        parser.add_argument('--skip_compiler_install', action="store_true", default=False,
+                            help="Skips the installation of the default compiler")
```
Long options usually use dashes as word separator. What do minifi options get passed to?

martinzink replied:
We pass that directly to the first cmake command that parses the minifioptions cmake file

szaszm replied:
I don't really understand the answer, so I'll try to rephrase my question: We should use `--minifi-options` instead of `--minifi_options`, `--cmake-options` instead of `--cmake_options`, and `--skip-compiler-install` instead of `--skip_compiler_install`, because in command line long options, dashes are used as the word separator in most cases.
I'm not sure how cmake fits here? These long option names are parsed by python, aren't they?

martinzink replied:
sry misunderstood the question, I see your point I changed that in https://github.com/apache/nifi-minifi-cpp/pull/1681/commits/faabed2da474a73ae6f3761eeccf128cd09426e3

szaszm commented:
I tried running this on windows 10, in a VS2022 developer command prompt. It prompted me to install the vs2022 build tools, which shouldn't be necessary with a full VS installation, and failed to run when I answered with no.
![Screenshot_win10_2024-02-06_13:46:31](https://github.com/apache/nifi-minifi-cpp/assets/1170582/3909d08f-6a3f-486c-81d7-fa3ea2f318b8)

fgerlits commented on bootstrap/py_bootstrap.sh:
To accomodate users who don't read instructions, we should first check if `python` and `python-venv` are installed, and tell the user to install them if they aren't.

Also, if `python` is installed but `python-venv` isn't, the script creates a broken `venv` directory, so it won't work even after `python-venv` is installed later.  The `venv` directory should be removed (or not created) in this case.

fgerlits commented on bootstrap/cli.py:
After a step has finished in `Step by step` mode, is it possible to return to the `Step by step` menu instead of the main menu?

martinzink commented:
> I tried running this on windows 10, in a VS2022 developer command prompt. It prompted me to install the vs2022 build tools, which shouldn't be necessary with a full VS installation, and failed to run when I answered with no. ![Screenshot_win10_2024-02-06_13:46:31](https://private-user-images.githubusercontent.com/1170582/302628730-3909d08f-6a3f-486c-81d7-fa3ea2f318b8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgwMDg2OTcsIm5iZiI6MTcwODAwODM5NywicGF0aCI6Ii8xMTcwNTgyLzMwMjYyODczMC0zOTA5ZDA4Zi02YTNmLTQ4NmMtODFkNy1mYTNlYTJmMzE4YjgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIxNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMTVUMTQ0NjM3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTRkNjNhZjYwMWRjMmM5NDE1NWU0NzM5YmZmMjViYTZiZGY4MTgyMjQ5YWIzYjU0ZjQ1NjA5NzkxMWZmMDVhZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.lbhbstMNMHrwzuSiQNB7rmhJGs-q3z4sqAkj6WsJJIg)

Im curious does vswhere find this installation, and do you have the required packages installed?
In theory if vswhere finds your installation and it satisfies the requirements (C++ with VC.ATL support) it shouldnt prompt for an install, but if it doesnt find it, it wont work. Because it needs a valid installation.

fgerlits reviewed (approved):
Do we want to remove `/bootstrap.sh`?  Or are we keeping it for a while until possible kinks are ironed out with the new boostrap?

lordgamez commented:
> > I tried running this on windows 10, in a VS2022 developer command prompt. It prompted me to install the vs2022 build tools, which shouldn't be necessary with a full VS installation, and failed to run when I answered with no. ...
> 
> Im curious does vswhere find this installation, and do you have the required packages installed? In theory if vswhere finds your installation and it satisfies the requirements (C++ with VC.ATL support) it shouldnt prompt for an install, but if it doesnt find it, it wont work. Because it needs a valid installation.

I had a similar experience at first, then after I cancelled the vs buildtools install, uninstalled the build tools from the VS installer, and I saw my VS installation was out of date, so updated it just in case, then after retrying the bootstrapper it didn't want to try installing the build tools. Not sure if the VS update solved the issue or it was something else.

lordgamez commented:
An other issue I had with the minificontroller build was this on Windows:

```
FAILED: bin/minificontroller.exe
cmd.exe /C "cd . && "C:\Program Files\CMake\bin\cmake.exe" -E vs_link_exe --intdir=controller\CMakeFiles\minificontroller.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\mt.exe --manifests  -- C:\PROGRA~1\MICROS~1\2022\COMMUN~1\VC\Tools\MSVC\1439~1.335\bin\Hostx64\x64\link.exe /nologo controller\CMakeFiles\minificontroller.dir\MiNiFiController.cpp.obj controller\CMakeFiles\minificontroller.dir\Controller.cpp.obj controller\CMakeFiles\minificontroller.dir\__\minifi_main\MainHelper.cpp.obj controller\CMakeFiles\minificontroller.dir\versioninfo.rc.res  /out:bin\minificontroller.exe /implib:controller\minificontroller.lib /pdb:bin\minificontroller.pdb /version:0.0 /machine:x64 /INCREMENTAL:NO /subsystem:console  bin\core-minifi.lib  thirdparty\yaml-cpp-install\lib\yaml-cpp.lib  thirdparty\zlib-install\lib\zlibstatic.lib  _deps\spdlog-build\spdlog.lib  _deps\fmt-build\fmt.lib  thirdparty\libsodium-install\lib\sodium.lib  date-tz.lib  thirdparty\openssl-install\lib\libssl.lib  thirdparty\openssl-install\lib\libcrypto.lib  crypt32.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cmd.exe /C "cd /D D:\Projects\nifi-minifi-cpp\build\controller && cat D:/Projects/nifi-minifi-cpp/build/all.log""
'cat' is not recognized as an internal or external command,
operable program or batch file.
```

martinzink commented:
> Do we want to remove `/bootstrap.sh`? Or are we keeping it for a while until possible kinks are ironed out with the new bootstrap?

I think the latter would be better.

lordgamez commented:
I created a clean new Windows VM and tried out the new python bootstrap to set up the environment. When it installed the dependencies and got into the build menu I exited the bootstrap, and tried if the build can be still run using the win_build_vs.bat script. The configuration was successful and the build started, but failed with the following issues:

```
Performing configure step for 'openssl-external'
'perl' is not recognized as an internal or external command,
operable program or batch file.
```
```
Failure!  build file wasn't produced.
Please read INSTALL.md and associated NOTES-* files.  You may also have to
look over your available compiler tool chain or change your configuration.

NASM not found - make sure it's installed and available on %PATH%
```
Are the perl and NASM dependencies installed in build time using the bootstrap build command, or are they missing somehow? After I installed these 2 dependencies manually the build continued.

martinzink commented:
> I created a clean new Windows VM and tried out the new python bootstrap to set up the environment. When it installed the dependencies and got into the build menu I exited the bootstrap, and tried if the build can be still run using the win_build_vs.bat script. The configuration was successful and the build started, but failed with the following issues:
> 
> ```
> Performing configure step for 'openssl-external'
> 'perl' is not recognized as an internal or external command,
> operable program or batch file.
> ```
> 
> ```
> Failure!  build file wasn't produced.
> Please read INSTALL.md and associated NOTES-* files.  You may also have to
> look over your available compiler tool chain or change your configuration.
> 
> NASM not found - make sure it's installed and available on %PATH%
> ```
> 
> Are the perl and NASM dependencies installed in build time using the bootstrap build command, or are they missing somehow? After I installed these 2 dependencies manually the build continued.

They are installed but the choco installer doesnt add them to path. I didnt want to modify the path because strawberry ssl and nasm can mess up the path, so they are only added temporaraly to path during the bootstrap process.

martinzink commented:
> An other issue I had with the minificontroller build was this on Windows:
> 
> ```
> FAILED: bin/minificontroller.exe
> cmd.exe /C "cd . && "C:\Program Files\CMake\bin\cmake.exe" -E vs_link_exe --intdir=controller\CMakeFiles\minificontroller.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\mt.exe --manifests  -- C:\PROGRA~1\MICROS~1\2022\COMMUN~1\VC\Tools\MSVC\1439~1.335\bin\Hostx64\x64\link.exe /nologo controller\CMakeFiles\minificontroller.dir\MiNiFiController.cpp.obj controller\CMakeFiles\minificontroller.dir\Controller.cpp.obj controller\CMakeFiles\minificontroller.dir\__\minifi_main\MainHelper.cpp.obj controller\CMakeFiles\minificontroller.dir\versioninfo.rc.res  /out:bin\minificontroller.exe /implib:controller\minificontroller.lib /pdb:bin\minificontroller.pdb /version:0.0 /machine:x64 /INCREMENTAL:NO /subsystem:console  bin\core-minifi.lib  thirdparty\yaml-cpp-install\lib\yaml-cpp.lib  thirdparty\zlib-install\lib\zlibstatic.lib  _deps\spdlog-build\spdlog.lib  _deps\fmt-build\fmt.lib  thirdparty\libsodium-install\lib\sodium.lib  date-tz.lib  thirdparty\openssl-install\lib\libssl.lib  thirdparty\openssl-install\lib\libcrypto.lib  crypt32.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cmd.exe /C "cd /D D:\Projects\nifi-minifi-cpp\build\controller && cat D:/Projects/nifi-minifi-cpp/build/all.log""
> 'cat' is not recognized as an internal or external command,
> operable program or batch file.
> ```

I think this was caused by an old post build step, removed it in 
https://github.com/apache/nifi-minifi-cpp/pull/1681/commits/9f39511a1a991454311992cbd5dbf4e128a27051

szaszm commented on extensions/grafana-loki/tests/PushGrafanaLokiRESTTest.cpp:
thx for fixing this, I also ran into this issue a few weeks ago.

szaszm commented on .github/workflows/ci.yml:
```diff
@@ -75,12 +131,72 @@ jobs:
+        -DENABLE_BZIP2=ON
+        -DENABLE_CIVET=ON
+        -DENABLE_COAP=OFF
+        -DENABLE_CONTROLLER=OFF
```
Now that you fixed its build, I think we should enable the controller even on Windows.

martinzink replied:
good idea, https://github.com/apache/nifi-minifi-cpp/pull/1681/commits/48ebdcec251fc04221d0cbcf750bc261eed9409f

szaszm commented on Windows.md:
```diff
@@ -120,7 +120,7 @@ A basic working CMake configuration can be inferred from the `win_build_vs.bat`.
 mkdir build
 cd build
-cmake -G "Visual Studio 17 2022" -A x64 -DINSTALLER_MERGE_MODULES=OFF -DTEST_CUSTOM_WEL_PROVIDER=OFF -DENABLE_SQL=OFF -DUSE_REAL_ODBC_TEST_DRIVER=OFF -DCMAKE_BUILD_TYPE_INIT=Release -DCMAKE_BUILD_TYPE=Release -DWIN32=WIN32 -DENABLE_LIBRDKAFKA=OFF -DENABLE_JNI=OFF -DOPENSSL_OFF=OFF -DENABLE_COAP=OFF -DENABLE_AWS=OFF -DENABLE_PDH= -DENABLE_AZURE=OFF -DENABLE_SFTP=OFF -DENABLE_SPLUNK= -DENABLE_GCP=  ...
+cmake -G "Visual Studio 16 2022" -A x64 -DINSTALLER_MERGE_MODULES=OFF -DTEST_CUSTOM_WEL_PROVIDER=OFF -DENABLE_SQL=OFF -DMINIFI_USE_REAL_ODBC_TEST_DRIVER=OFF -DCMAKE_BUILD_TYPE_INIT=Release -DCMAKE_BUILD_TYPE=Release -DWIN32=WIN32 -DENABLE_LIBRDKAFKA=OFF -DENABLE_JNI=OFF -DMINIFI_OPENSSL=ON -DENABLE_COAP=OFF -DENABLE_AWS=OFF -DENABLE_PDH= -DENABLE_AZURE=OFF -DENABLE_SFTP=OFF -DENABLE_SPLUNK= -DENA ...
```
On my machine, there's `Visual Studio 16 2019` and `Visual Studio 17 2022`, but no `Visual Studio 16`. Was this change made by accident?

szaszm commented on bootstrap/py_bootstrap.bat:
```diff
@@ -0,0 +1,29 @@
+) else (
+    echo Creating virtualenv
+    python -m venv "%VENV_DIR%"
+    call "%VENV_DIR%\Scripts\activate"
```
Aren't we supposed to call activate.bat here? Why does this even work?

martinzink replied:
Windows allows referencing files without explicit extensions in some scenarios. Made it explicit in [replace activate with activate.bat](https://github.com/apache/nifi-minifi-cpp/pull/1681/commits/833065382bd84716d239c0f2b7871ae8a66dd5da)

szaszm replied:
I was especially confused, because there is an activate file in the same location without an extension, but it's a bash script. Thx for changing it.

szaszm commented on bootstrap/package_manager.py:
```diff
@@ -0,0 +1,259 @@
+
+
+def _get_vs_dev_cmd_path() -> str:
+    vswhere_results = subprocess.run(
+        "vswhere -products * -property installationPath -requires Microsoft.VisualStudio.Component.VC.ATL",
+        capture_output=True)
```
`vswhere` is not found on my machine. It's present in the VS installer directory, but not in the path. When I installed it with scoop, this could use that version, but not the one that's automatically installed with visual studio. 
![image](https://github.com/apache/nifi-minifi-cpp/assets/1170582/f44142ea-308e-4710-94d0-7bb39d2b8ebe)


This script is supposed to work even without VS installed, right? Where does it take vswhere from in that case?

martinzink replied:
Yeah the script uses the package manager to tell if a required package/lib/exe is installed or not.
Maybe instead of the package manager we could use where.exe to tell if vswhere is installed or not. Probably this is the only instance where this would be useful (and vswhere is a tiny exe), so not sure if it worth the hassle to add more complication to the code. What do you think?

szaszm replied:
I think it's worth having an exception for vswhere.exe if it's already installed with VS. According to microsoft/vswhere readme, it's always at `%ProgramFiles(x86)%\Microsoft Visual Studio\Installer\vswhere.exe` when it's installed.
There's also a problem with it finding VS2019 instead of VS2022, as I have both installed to be able to compile older versions of minifi.

martinzink replied:
I've addressed these issues in https://github.com/apache/nifi-minifi-cpp/pull/1681/commits/46b3ceb968812ee09fdd6eca70aba37b967a2f25

szaszm commented:
In the same Windows environment I had issues with `vswhere`, I also had issue with the compiler command line after installing vswhere with scoop and refusing the offered choco install.
```
-- The C compiler identification is unknown
-- The CXX compiler identification is unknown
CMake Error at CMakeLists.txt:25 (project):
  The CMAKE_C_COMPILER:

    C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.38.33130/bin/Hostx64/x64/cl.exe

  is not a full path to an existing compiler tool.

  Tell CMake where to find the compiler by setting either the environment
  variable "CC" or the CMake cache entry CMAKE_C_COMPILER to the full path to
  the compiler, or to the compiler name if it is in the PATH.


CMake Error at CMakeLists.txt:25 (project):
  The CMAKE_CXX_COMPILER:

    C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.38.33130/bin/Hostx64/x64/cl.exe

  is not a full path to an existing compiler tool.

  Tell CMake where to find the compiler by setting either the environment
  variable "CXX" or the CMake cache entry CMAKE_CXX_COMPILER to the full path
  to the compiler, or to the compiler name if it is in the PATH.


-- Configuring incomplete, errors occurred!
```

martinzink commented:
> In the same Windows environment I had issues with `vswhere`, I also had issue with the compiler command line after installing vswhere with scoop and refusing the offered choco install.
> 
> ```
> -- The C compiler identification is unknown
> -- The CXX compiler identification is unknown
> CMake Error at CMakeLists.txt:25 (project):
>   The CMAKE_C_COMPILER:
> 
>     C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.38.33130/bin/Hostx64/x64/cl.exe
> 
>   is not a full path to an existing compiler tool.
> 
>   Tell CMake where to find the compiler by setting either the environment
>   variable "CC" or the CMake cache entry CMAKE_C_COMPILER to the full path to
>   the compiler, or to the compiler name if it is in the PATH.
> 
> 
> CMake Error at CMakeLists.txt:25 (project):
>   The CMAKE_CXX_COMPILER:
> 
>     C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.38.33130/bin/Hostx64/x64/cl.exe
> 
>   is not a full path to an existing compiler tool.
> 
>   Tell CMake where to find the compiler by setting either the environment
>   variable "CXX" or the CMake cache entry CMAKE_CXX_COMPILER to the full path
>   to the compiler, or to the compiler name if it is in the PATH.
> 
> 
> -- Configuring incomplete, errors occurred!
> ```

This looks interesting 🤔 , does the qouted path look correct? Is it the same that vswhere returns?

szaszm commented on Windows.md:
````diff
@@ -120,9 +120,22 @@ A basic working CMake configuration can be inferred from the `win_build_vs.bat`.
 ctest -C Release
 ```
+
+## Python based bootstrapping (recommended)
````
This could go on top, and the old instructions could go under another heading, like `## Alternative: Manual bootstrapping (advanced)`

martinzink replied:
https://github.com/apache/nifi-minifi-cpp/pull/1681/commits/194c8f3ab46266f43fbdf314e96759679b83b84b

szaszm commented on bootstrap/py_bootstrap.bat:
```diff
@@ -0,0 +1,29 @@
+)
+
+REM Check if venv module is available
+python -m venv --help > nul 2>&1
```
You could just remove the separate check step and handle the error when calling it normally below.

martinzink replied:
good idea, https://github.com/apache/nifi-minifi-cpp/pull/1681/commits/194c8f3ab46266f43fbdf314e96759679b83b84b

szaszm commented on bootstrap/py_bootstrap.sh:
```diff
@@ -0,0 +1,27 @@
+    exit 1
+fi
+
+# Check if virtualenv is installed
+if ! command -v python3 -m venv --help &>/dev/null; then
+    echo "virtualenv is not installed"
+    exit 1
+fi
+
+SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+VENV_DIR="$SCRIPT_DIR/venv"
+
+if [ -d "$VENV_DIR" ]; then
+    source "$VENV_DIR/bin/activate"
+else
+    echo "Creating virtualenv"
+    python3 -m venv "$VENV_DIR"
+    source "$VENV_DIR/bin/activate"
+    pip install -r "$SCRIPT_DIR/requirements.txt"
+fi
```
```suggestion
SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
VENV_DIR="$SCRIPT_DIR/venv"

set -e

if [ -d "$VENV_DIR" ]; then
    source "$VENV_DIR/bin/activate"
else
    echo "Creating virtualenv"
    if ! python3 -m venv "$VENV_DIR"; then
        echo "Creating virtualenv failed. Is venv installed?"
        exit 1
    fi
    source "$VENV_DIR/bin/activate"
    pip install -r "$SCRIPT_DIR/requirements.txt"
fi
```

szaszm commented on Windows.md:
I think most of the file should go under manual bootstrap and build, including Building with Visual Studio, and Building via the build script. The first section about Requirements should also be updated, to recommend the python bootstrap instead of the win_build_vs.bat script.

martinzink replied:
Good idea, https://github.com/apache/nifi-minifi-cpp/pull/1681/commits/194c8f3ab46266f43fbdf314e96759679b83b84b
We will probably need a followup PR anyways (removing the win_build_bat maybe even bootstrap.sh?) And fix the documentation
